File: plugin/src/main/java/com/stratio/cassandra/lucene/search/condition/builder/ConditionBuilder.java
/*
 * Licensed to STRATIO (C) under one or more contributor license agreements.
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.  The STRATIO (C) licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.stratio.cassandra.lucene.search.condition.builder;

import com.stratio.cassandra.lucene.search.condition.Condition;
import com.stratio.cassandra.lucene.util.Builder;
import org.codehaus.jackson.annotate.JsonProperty;
import org.codehaus.jackson.annotate.JsonSubTypes;
import org.codehaus.jackson.annotate.JsonTypeInfo;

/**
 * {@link Builder} for creating new {@link Condition}s.
 *
 * @param <T> The type of the {@link Condition} to be created.
 * @param <K> The specific type of the {@link ConditionBuilder}.
 * @author Andres de la Pena {@literal <adelapena@stratio.com>}
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({@JsonSubTypes.Type(value = AllConditionBuilder.class, name = "all"),
               @JsonSubTypes.Type(value = BitemporalConditionBuilder.class, name = "bitemporal"),
               @JsonSubTypes.Type(value = BooleanConditionBuilder.class, name = "boolean"),
               @JsonSubTypes.Type(value = ContainsConditionBuilder.class, name = "contains"),
               @JsonSubTypes.Type(value = DateRangeConditionBuilder.class, name = "date_range"),
               @JsonSubTypes.Type(value = FuzzyConditionBuilder.class, name = "fuzzy"),
               @JsonSubTypes.Type(value = GeoDistanceConditionBuilder.class, name = "geo_distance"),
               @JsonSubTypes.Type(value = GeoBBoxConditionBuilder.class, name = "geo_bbox"),
               @JsonSubTypes.Type(value = GeoShapeConditionBuilder.class, name = "geo_shape"),
               @JsonSubTypes.Type(value = LuceneConditionBuilder.class, name = "lucene"),
               @JsonSubTypes.Type(value = MatchConditionBuilder.class, name = "match"),
               @JsonSubTypes.Type(value = NoneConditionBuilder.class, name = "none"),
               @JsonSubTypes.Type(value = PhraseConditionBuilder.class, name = "phrase"),
               @JsonSubTypes.Type(value = PrefixConditionBuilder.class, name = "prefix"),
               @JsonSubTypes.Type(value = RangeConditionBuilder.class, name = "range"),
               @JsonSubTypes.Type(value = RegexpConditionBuilder.class, name = "regexp"),
               @JsonSubTypes.Type(value = WildcardConditionBuilder.class, name = "wildcard")})
public abstract class ConditionBuilder<T extends Condition, K extends ConditionBuilder<T, K>> implements Builder<T> {

    /** The boost for the {@link Condition} to be built. */
    @JsonProperty("boost")
    protected Float boost;

    /**
     * Sets the boost for the {@link Condition} to be built. Documents matching this condition will (in addition to the
     * normal weightings) have their score multiplied by {@code boost}. If {@code null}, then {@link
     * Condition#DEFAULT_BOOST}
     *
     * @param boost the boost
     * @return this with the specified boost
     */
    @SuppressWarnings("unchecked")
    public K boost(float boost) {
        this.boost = boost;
        return (K) this;
    }

    /**
     * Sets the boost for the {@link Condition} to be built. Documents matching this condition will (in addition to the
     * normal weightings) have their score multiplied by {@code boost}. If {@code null}, then {@link
     * Condition#DEFAULT_BOOST}
     *
     * @param boost he boost
     * @return this with the specified boost
     */
    @SuppressWarnings("unchecked")
    public K boost(Number boost) {
        this.boost = boost == null ? null : boost.floatValue();
        return (K) this;
    }

    /**
     * Returns the {@link Condition} represented by this builder.
     *
     * @return a new condition
     */
    @Override
    public abstract T build();
}

File: plugin/src/main/scala/com/stratio/cassandra/lucene/search/condition/builder/ConditionBuilder.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.search.condition.builder

import com.fasterxml.jackson.annotation.{JsonProperty, JsonSubTypes, JsonTypeInfo}
import com.stratio.cassandra.lucene.common.Builder
import com.stratio.cassandra.lucene.search.condition.Condition

/**
 * [[Builder]] for creating new [[Condition]]s.
 *
  * @tparam T The type of the [[Condition]] to be created.
 * @tparam K  The specific type of the [[ConditionBuilder]].
 * @author Andres de la Pena `adelapena@stratio.com`
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type", defaultImpl = classOf[BooleanConditionBuilder])
@JsonSubTypes(Array(
                new JsonSubTypes.Type(value = classOf[AllConditionBuilder], name = "all"),
                new JsonSubTypes.Type(value = classOf[BitemporalConditionBuilder], name = "bitemporal"),
                new JsonSubTypes.Type(value = classOf[BooleanConditionBuilder], name = "boolean"),
                new JsonSubTypes.Type(value = classOf[ContainsConditionBuilder], name = "contains"),
               new JsonSubTypes.Type(value = classOf[DateRangeConditionBuilder], name = "date_range"),
               new JsonSubTypes.Type(value = classOf[FuzzyConditionBuilder], name = "fuzzy"),
               new JsonSubTypes.Type(value = classOf[GeoDistanceConditionBuilder], name = "geo_distance"),
               new JsonSubTypes.Type(value = classOf[GeoBBoxConditionBuilder], name = "geo_bbox"),
               new JsonSubTypes.Type(value = classOf[GeoShapeConditionBuilder], name = "geo_shape"),
               new JsonSubTypes.Type(value = classOf[LuceneConditionBuilder], name = "lucene"),
               new JsonSubTypes.Type(value = classOf[MatchConditionBuilder], name = "match"),
               new JsonSubTypes.Type(value = classOf[NoneConditionBuilder], name = "none"),
               new JsonSubTypes.Type(value = classOf[PhraseConditionBuilder], name = "phrase"),
               new JsonSubTypes.Type(value = classOf[PrefixConditionBuilder], name = "prefix"),
               new JsonSubTypes.Type(value = classOf[RangeConditionBuilder], name = "range"),
               new JsonSubTypes.Type(value = classOf[RegexpConditionBuilder], name = "regexp"),
               new JsonSubTypes.Type(value = classOf[WildcardConditionBuilder], name = "wildcard")))
abstract class ConditionBuilder[T >: Condition, K >: ConditionBuilder[T,K]] extends Builder[T] {

    var boost : Float = 1.0f
    /**
     * Returns the [[Condition]] represented by this builder.
     *
     * @return a new condition
     */
    def build : T

    @JsonProperty("boost")
    def boost(_boost : Float ) : K = {
      this.boost=_boost
      this.asInstanceOf[K]
    }
}
